feat: Adapt Dialog to use Drawers on mobile

gramo37 opened the pull request:
## What does this PR do?

Fixes the below issue.
[CAL-2978] Adapt Dialog to use Drawers on modals #13338

fixes #13338

Please refer the below video and let me know if the solution is as expected:

Screen recordings of Production build -> 
https://www.loom.com/share/50f6857055ec4bb0848f7d5d287fc52c?sid=2015e908-3942-4a0c-b469-9f64358d0bbd

https://www.loom.com/share/51e98373a2424d90b09cdf528af8eea3?t=51&sid=295c5ed6-5d99-408f-9c60-e7c6b0ac54ed

## Type of change

- New feature (non-breaking change which adds functionality)

## How should this be tested?

- Open any dialog.
- Adjust the screen width to observe the changes.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=13633) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=13633) before we can accept your contribution.<br/><hr/>**prasanna** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=13633) it.</sub>

sean-brydon commented:
@ciaranha - can you checkout preview for this PR - code looks good on my end.

keithwillcode reviewed (changes requested):
These changes seem to have broken unit and E2E tests.

ciaranha commented:
> @ciaranha - can you checkout preview for this PR - code looks good on my end. 

These look great. Only slight nit is id probably reduce the padding inside the bottom sheet on mobile by at least 8px.

gramo37 commented:
Hi @keithwillcode, I've addressed the issues causing the test cases to fail. Here's what I found:

1)  There was no window.matchMedia in the testing environment (TypeError: window.matchMedia is not a function)
2) DialogOverlay must be utilized within Dialog.

I have made changes for that, and now the tests for the Dialog component are passing smoothly.

![passed_testcase](https://github.com/calcom/cal.com/assets/69106022/868341bf-18be-4ec7-82d8-8c16293ee566)

I have also added the change mentioned by @ciaranha. Reduced the padding for Footer in the mobile view.

Udit-takkar commented on packages/ui/components/dialog/Dialog.tsx:
```diff
@@ -175,6 +232,22 @@ DialogContent.displayName = "DialogContent";
 // export const DialogClose = DialogPrimitive.Close;
 
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function DialogCloseWrapper({ children, ...props }: any) {
```
we don't prefer to use 'any' type in the codebase

Udit-takkar replied:
can you replace it with actual type?

gramo37 replied:
I have replaced it. Thanks for reviewing.

Udit-takkar commented on packages/ui/components/dialog/Dialog.tsx:
```diff
@@ -161,8 +217,9 @@ export function DialogFooter(props: { children: ReactNode; className?: string; s
         className={classNames(
-          "flex justify-end space-x-2 pb-4 pt-4 rtl:space-x-reverse",
-          !props.showDivider && "pb-8"
+          "flex justify-end space-x-2 rtl:space-x-reverse",
+          isMobile ? "pb-2 pt-2" : "pb-4 pt-4",
+          !props.showDivider && (isMobile ? "pb-6" : "pb-8")
```
you could use breakpoints here instead of isMobile. https://tailwindcss.com/docs/responsive-design

gramo37 replied:
Also added this.

Udit-takkar commented on packages/ui/components/dialog/Dialog.tsx:
```diff
@@ -68,8 +71,76 @@ export function Dialog(props: DialogProps) {
+}
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+function DialogContentWrapper({ children, ...props }: any) {
```
same here

CarinaWolli reviewed (changes requested):
I noticed one thing: On my desktop browser everything looks great and scrolling works when the content gets too long. However when I tested it on my phone the scrolling didn't work. Any idea why that is? (tested chrome and safari on iphone)

NIT: can we fix the bottom rounded corners?
![Screenshot 2024-02-19 at 7 43 44 PM](https://github.com/calcom/cal.com/assets/30310907/08d0a757-fe26-4e2b-8d59-47de1e4ee7cf)

CarinaWolli reviewed (changes requested):
Scrolling works now 👍

In general, drawers look good. But I found quite a few that have issues. I didn't check all dialogs, but here are some issues I found: 

- Unwanted padding:
![Screenshot 2024-02-21 at 12 08 42 PM](https://github.com/calcom/cal.com/assets/30310907/59402048-9238-487c-b0c6-46327547f908)

- Dropdown cut-off (add workflow action & set priority for round robin host) 
![Screenshot 2024-02-21 at 12 09 41 PM](https://github.com/calcom/cal.com/assets/30310907/d8cf2cc7-d699-4c72-846e-4a37fa7d3316)
![Screenshot 2024-02-21 at 12 23 00 PM](https://github.com/calcom/cal.com/assets/30310907/3c3cc4f7-04bb-421e-89aa-de67f220f52a)

- Heading centered:
![Screenshot 2024-02-21 at 12 19 49 PM](https://github.com/calcom/cal.com/assets/30310907/cd158307-78d4-42ee-9215-3464801a61bf)

- Enabling two factor authentication doesn't open the drawer at all

gramo37 commented:
> First 2 issues (Unwanted padding, Dropdown cut-off) have been fixed and I have pushed the changes for them.
> 
> ![image](https://private-user-images.githubusercontent.com/69106022/307503561-d8409d62-1104-4904-89c4-7563163be5e2.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjIzNzksIm5iZiI6MTcwODg2MjA3OSwicGF0aCI6Ii82OTEwNjAyMi8zMDc1MDM1NjEtZDg0MDlkNjItMTEwNC00OTA0LTg5YzQtNzU2MzE2M2JlNWUyLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExNTQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTljMWQyODM0ZDI5OWFiMWE3ZTA3NGMwOGEzMTNiOTBkYmM4NmI5Mzc4YmIwODE1MjhhNWIxMjEyNDIzMmJkMWEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.pt4HumjOMl7_trIJQQk3f5UXirIHGpWgIRtDTWEMY14)
> 
> However, for the third one, I will have to make changes in EditLocationDialog.tsx. As the heading is centered in the Dialog as well.
> 
> ![image](https://private-user-images.githubusercontent.com/69106022/307503681-5073c092-57a7-46e5-b9f7-203f0a1c50b0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjIzNzksIm5iZiI6MTcwODg2MjA3OSwicGF0aCI6Ii82OTEwNjAyMi8zMDc1MDM2ODEtNTA3M2MwOTItNTdhNy00NmU1LWI5ZjctMjAzZjBhMWM1MGIwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExNTQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWMzMmNjOGVhYWM3OGY4NGY1NmYyMzM2NzdkZDNjYzRlMjA2YjQ3ZGU3YzFlY2IwMDFlNTRjMTViZTVmMzU2OTgmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.R01KyUBh-0Zq-oo068xxEklBQ6G551WBW2NwFRIjGrI)
> 
> For the last issue, when I passed "setEnableModalOpen" instead of "() => setEnableModalOpen(!enableModalOpen)" to onOpenChange prop the problem was solved (for both Desktop and mobile view). But I am thinking if it is a correct solution or not.
> 
> ![image](https://private-user-images.githubusercontent.com/69106022/307503973-824e67c5-b6db-45a8-a9b9-e56c73046d7f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjIzNzksIm5iZiI6MTcwODg2MjA3OSwicGF0aCI6Ii82OTEwNjAyMi8zMDc1MDM5NzMtODI0ZTY3YzUtYjZkYi00NWE4LWE5YjktZTU2YzczMDQ2ZDdmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExNTQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ3NmI3M2E0ZjNiYTM5NmM3NDEzMzhlZDQ5YmVkYjljNjQwOWE5MWVmYWNiOWViM2E3NTk3Y2Y2NjRjZjNlMzUmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.E4R2uS_55ckK39jWXXt76Dm8KAuvOPAK21qlxtcmdDw)
> 
> ![image](https://private-user-images.githubusercontent.com/69106022/307504207-b8219538-271c-4e96-ba8f-53369a9d2865.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg4NjIzNzksIm5iZiI6MTcwODg2MjA3OSwicGF0aCI6Ii82OTEwNjAyMi8zMDc1MDQyMDctYjgyMTk1MzgtMjcxYy00ZTk2LWJhOGYtNTMzNjlhOWQyODY1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAyMjUlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMjI1VDExNTQzOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTgyNjhlYzY1OTdmZDM3ODkzNmRmYTExNDQzYjMwNThkOWUzNGRmOTRlZTlkYjBlYjIwNGZkY2VkZWY4OGIxN2MmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.S3vPqz-D4I9D0mvx5op_gelKiRU1xXNMIzcvY-vzp4w)
> 
> Please let me know if I should push the last 2 changes.

I found a similar case in conferencing.tsx (DisconnectIntegrationModal), which could be fixed using a similar approach. I think there might me some issue in the vaul library which is opening and closing the Drawer again and again in some Modals. To solve this, will have to check all Dialogs and solve the errors in a similar manner or add a prop to Dialog named "useModalForMobile". So, when it is set to true the Dialog will be used instead of drawer for such cases.

CarinaWolli commented:
> Please let me know if I should push the last 2 changes.

Sounds good to me, let's push them

> I found a similar case in conferencing.tsx (DisconnectIntegrationModal), which could be fixed using a similar approach. I think there might me some issue in the vaul library which is opening and closing the Drawer again and again in some Modals. To solve this, will have to check all Dialogs and solve the errors in a similar manner or add a prop to Dialog named "useDialogForMobile". So, when it is set to true the Dialog will be used instead of drawer for such cases.

We will need to go and test all modals to avoid breaking anything. I would definitely go for fixing the issues and using Drawers everywhere, if the issues are more complex we could temporary disable the Drawer for specific modals but I would like to avoid that if possible

gramo37 commented:
I have checked as many problems found in Dialogs. Some of them have been fixed but can't find a solution for some. Therefore, I have added the prop useDialogForMobile to Dialog.tsx and used it wherever required.

Problems
1) CSS problem in  EditLocationDialog.tsx. Solution -> Aligned the title to the left (CSS problem)  
2) TwoFactorAuthentication. Solution -> Minor change in onOpenChange prop 
3) Removing Installed Apps had a problem while opening the Drawer.  Solution -> Minor change in open and onOpenChange prop
4) Problem in Create customized event names (Steps to reproduce: edit event type -> advanced tab). Solution -> Used the useDialogForMobile prop
5) Reschedule booking not opening drawer after selecting a time slot. Solution -> Used the useDialogForMobile prop
   

> > Please let me know if I should push the last 2 changes.
> 
> Sounds good to me, let's push them
> 
> > I found a similar case in conferencing.tsx (DisconnectIntegrationModal), which could be fixed using a similar approach. I think there might me some issue in the vaul library which is opening and closing the Drawer again and again in some Modals. To solve this, will have to check all Dialogs and solve the errors in a similar manner or add a prop to Dialog named "useDialogForMobile". So, when it is set to true the Dialog will be used instead of drawer for such cases.
> 
> We will need to go and test all modals to avoid breaking anything. I would definitely go for fixing the issues and using Drawers everywhere, if the issues are more complex we could temporary disable the Drawer for specific modals but I would like to avoid that if possible

CarinaWolli commented:
Moving this to draft because there are still some issues. Thank you for the PR, we will come back to it when we have the bandwidth to carefully test this so we can make sure to not break anything 🙏

keithwillcode commented:
@anikdhabal or @Amit91848 one of you able to get this tested and pushed through? 🙏🏼

anikdhabal commented:
> @anikdhabal or @Amit91848 one of you able to get this tested and pushed through? 🙏🏼

sure 🙏🏼

anikdhabal reviewed (commented):
Hey @gramo37 could you pls fix the conflict?

anikdhabal reviewed (commented):
Not adapted in the Invite team member modal :-
![Screenshot 2024-06-29 171241](https://github.com/calcom/cal.com/assets/81948346/f2ee2d3c-4da8-46a8-a682-99b6320d53e6)

![Screenshot 2024-06-29 180140](https://github.com/calcom/cal.com/assets/81948346/80a71e36-d2b2-4394-bc12-4e08686137c5)

anikdhabal reviewed (commented):
Remove member modal is not opening and the screen has started flickering:-

https://github.com/calcom/cal.com/assets/81948346/d439d786-606e-4c44-bb8f-36c067c1a352

anikdhabal reviewed (commented):
See this weird behavior at the time of opening event-type-creation dialog:

https://github.com/calcom/cal.com/assets/81948346/7552e63e-f62d-4eff-9b40-6d8b96ea4d2a


- Same for the duplicate event type dialog.
- Same for the add question dialog

anikdhabal reviewed (commented):
Fix the padding:-
![Screenshot 2024-06-29 175102](https://github.com/calcom/cal.com/assets/81948346/92512ee5-b1c2-44d8-873f-c554ce77304d)

anikdhabal reviewed (changes requested):
Left some messages. Pls address them.

gramo37 commented:
> Hey @gramo37 could you pls fix the conflict?

I have resolved the merge conflict. Since many changes were required, I created a new pull request for it. Please refer to PR https://github.com/calcom/cal.com/pull/15632
